src/Raven.Server/Utils/CpuUsageExtensionPoint.cs

```diff
@@ -0,0 +1,263 @@
+                }
+
+                _process = null;
+            }
+        }
+
+
+        private async Task ReadProcess()
+        {
+            while (IsDisposed == false)
+            {
+                var nextLine = _process.StandardOutput.ReadLineAsync();
+
+                if (await nextLine.WaitWithTimeout(_timeout) == false)
+                {
+                    // here we missed an update, restart the process once
+                    HandleError($"The process didn't send information for {_timeout.TotalSeconds} seconds.");
+                    break;
+                }
+
+                var line = await nextLine;
+                if (HandleInfoReceived(line) == false)
+                {
+                    break;
+                }
+            }
+        }
+
+        private bool HandleInfoReceived(string data)
+        {
+            if (data == null)
+            {
+                HandleError("The output stream of the process has closed.");
+                return false;
+            }
+            using (_contextPool.AllocateOperationContext(out var context))
+            {
+                try
+                {
+                    var blittable = context.ReadForMemory(data, string.Empty);
+                    if (blittable.TryGet(nameof(ExtensionPointRawData.MachineCpuUsage), out double machineCpuUsage) == false)
+                    {
+                        HandleError($"Can't read {nameof(ExtensionPointRawData.MachineCpuUsage)} property from : \n{blittable}.");
+                        return false;
+                    }
+                    _data.MachineCpuUsage = machineCpuUsage;
+
+                    if (blittable.TryGet(nameof(ExtensionPointRawData.ActiveCores), out double activeCores) == false)
+                    {
+                        HandleError($"Can't read {nameof(ExtensionPointRawData.ActiveCores)} property from : \n{blittable}.");
+                        return false;
+                    }
+                    _data.ActiveCores = activeCores;
+                }
+                catch (Exception e)
+                {
+                    HandleError($"Unable to parse \"{data}\" to json", e);
+                    return false;
+                }
+            }
+
+            return true;
+        }
+
+        private void HandleError(string msg, Exception e = null)
+        {
+            try
+            {
+                if (_process.HasExited == false)
+                {
+                    _process.Kill();
+                }
+            }
+            catch (Exception)
+            {
+                // When the process terminating, killed or exited
+            }
+            finally
+            {
+                if (_didRestart == false)
+                {
+                    _didRestart = true;
+                    NotifyWarning($"{msg} \nTherefore the process will restart.", e);
+
+                    InitNewProcessHandle();
```

Comment:
what if errors are thrown here?

Reply:
@ayende The `InitNewProcessHandle` function just set the new process object properties and `Start` function handles itself